File: leetcode/103-BinaryTreeZigzagLevelOrderTraversal/103-BinaryTreeZigzagLevelOrderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
    if (!root) return []; 

    let queue = [root];
    let toRight = true;
    let ans = [];

    while(queue.length){
        let nodesCurrentLevel = queue.length;
        let nextQueue = [];
        let res = [];
        for(let i = 0; i < nodesCurrentLevel; i++){
            let node = queue[i];
            res.push(node.val);

            if(node.right) nextQueue.push(node.right);
            if(node.left) nextQueue.push(node.left);
        }

        if(toRight){
            res.reverse();
        }
        ans.push(res)
        queue = nextQueue
        toRight = !toRight;

    }

    return ans;
};

File: leetcode/103-BinaryTreeZigzagLevelOrderTraversal/103-BinaryTreeZigzagLevelOrderTraversal.py
from collections import deque
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root: return []
        
        queue = deque([root])
        final = []
        toRight = True
        
        while queue:
            nodesOnLevel = len(queue)
            ans = []
            for node in range(nodesOnLevel):
                curr = queue.popleft()
                ans.append(curr.val)
                
                if curr.left:
                    queue.append(curr.left)  # Left child first
                if curr.right: 
                    queue.append(curr.right)  # Right child second
                    
            if not toRight:
                ans.reverse()
                    
            final.append(ans)
            toRight = not toRight

        
        return final
        
        
